### ISSUE TITLE
Can't export app. Getting error: {"Message":"Bad Request","ExceptionMessage":"The directory is not empty.  ","ExceptionType":"System.IO.IOException",...

### ISSUE THREAD

**Issue by NetLS-LLC** [2sic/2sxc#2857]
**I'm submitting a ...**   
[x] bug report => search github for a similar issue before submitting (did that, didn't find anything)

**...about**   
[x] edit experience / UI
[x] admin experience UI
[x] Content Types or data management

**Current behavior**   
I can import apps but I can't export any single app. I'm getting this error message:

> {"Message":"Bad Request","ExceptionMessage":"The directory is not empty.
> ","ExceptionType":"System.IO.IOException","StackTrace":"   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound, WIN32_FIND_DATA& data)
>    at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
>    at ToSic.Eav.Apps.ImportExport.ZipExport.ExportApp(Boolean includeContentGroups, Boolean resetAppGuid) in C:\\Projects\\2sxc\\eav-server\\ToSic.Eav.Apps\\ImportExport\\ZipExport.cs:line 162
>    at ToSic.Sxc.WebApi.ImportExport.ExportApp.Export(Int32 zoneId, Int32 appId, Boolean includeContentGroups, Boolean resetAppGuid) in C:\\Projects\\2sxc\\2sxc\\Src\\Sxc\\ToSic.Sxc.WebApi\\ImportExport\\ExportApp.cs:line 117
>    at ToSic.Sxc.WebApi.Admin.AppControllerReal`1.Export(Int32 zoneId, Int32 appId, Boolean includeContentGroups, Boolean resetAppGuid) in C:\\Projects\\2sxc\\2sxc\\Src\\Sxc\\ToSic.Sxc.WebApi\\Admin\\AppControllerReal.cs:line 88
>    at lambda_method(Closure , Object , Object[] )
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_2.<GetExecutor>b__2(Object instance, Object[] methodParameters)
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__17`1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__17`1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__5.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__6.MoveNext()"}

**Expected behavior**   
It's basic functionality of 2sxc module. Apps need to be exportable...

**Instructions to Reproduce the Problem**
On my PC I'm just going to certain app settings, tab App, "Export App".

**Why change the behavior?**  
Someone could stumble upon the same issue one day...

**Your environment**  

* **2sxc version(s):** 14.07.03; 14.08 (it's still an issue in the latest version 14.08)  
* **Browser:** [Chrome Version 105.0.5195.127 (Official Build) (64-bit)]
* **DNN:** [9.10.02 ] 
* **Language:** [English]

**Anything you would like to add**
I tried both with first checkbox to include data and without it. I've tried to update 2sxc module from 14.07.03 to 14.08. Also I've cleared cache of browser, app itself, server cache via personaBar. Also I've tried to delete all the website files on my local, restore them from git, rebuild modules and themes that need to be rebuilt, press "upgrade app" for 2sxc module. Non of that helped.
When I googled it I didn't find anything.
And I don't have such folder "C:\\Projects" so I don't understand where did this path come from.
Also I'm not sure whether you'll be able to reproduce it since this bug is reproducable only on my PC. 2 of other developers have the same website deployed on their local and they don't have this issue.
Looks like I have 2 options: reinstall windows or fix some small thing that is broken. I have no idea what it is and I don't understand what's going on from the error message. Full windows reinstall would be very time consuming way to fix it.
Could you please help me with it?

**Comment by tvatavuk** [comment]
2sxc code is trying to delete the temporary folder that was created as part of the normal 2sxc app export procedure. This temporary folder is created by 2sxc code and it should be able to delete it, but for some external reason, it can't.

It could be because of an error on the disk or because the windows account used to run dnn does not have delete permission, or because antivirus or some other kind of protection is blocking the delete operation. You have to check and solve this on your system.

Temp folder is created in your `...\DesktopModules\ToSIC_SexyContent\_\`. This folder should be empty if all is ok. You are free to make it empty.

The path `C:\\Projects\\2sxc\\2sxc\\Src\\Sxc\\ToSic.Sxc.WebApi\\Admin\\AppControllerReal.cs` is from default 2sxc development environment, so it is not expected to be on other computers.

**Comment by NetLS-LLC** [comment]
Looks like the root cause of this issue is from [here](https://stackoverflow.com/a/47398010/7402089):
visual studio for some reason was blocking folders in that temporary folder and 2sxc module couldn't delete folders. All the files were deleted though, but some folders stayed empty in that temporary directory that you've mentioned.
Closing visual studio fixed it. Also I've noticed that very small app was successfully exported most of the time and big complex app was mostly generating that error. I guess it was just a chance of whether export and vs file system update from drive to RAM or cache were done at the same time or not.

Probably some loop with certain amount of repeated attempts, wrapped in try/catch and timeout in between could resolve this issue, cause it seems that windows explorer and visual studio are not using those files and folders all the time. Looks like they rather read it every now and then to keep it up to date in UI window.
